enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,518 @@
+  - "@mrunalp"
+creation-date: "2024-05-05"
+last-updated: "2024-06-12"
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - "https://issues.redhat.com/browse/CNV-28178"
+see-also:
+  - "https://github.com/kubernetes/enhancements/issues/2400"
+status: implementable
+---
+
+# Virtualization higher workload density
+
+## Summary
+
+Fit more workloads onto a give node - achieve a higher workload
+density - by overcommitting it's memory resources. Due to timeline
+needs a two-phased approach is considered.
+
+## Motivation
+
+Today, OpenShift Virtualization is reserving memory (`requests.memory`)
+according to the needs of the virtual machine itself. However, usually
+an application within the virtual machine does not utilize _all_ the
+memory _all_ the time. Instead, only _sometimes_ there are memory spikes
+within the virtual machine. Extrapolating this behavior to all virtual
+machines on a given node leads to the observation that _most often_ there
+is no memory pressure and a low memory utilization.
+Underutilized hardware resources are a cost factor to end-users.
+
+This proposal is about increasing the workload - virtual machine - density
+per node, in order to increase the memory utilization, in order to
+reduce the cost per virtual machine.
+
+> [!NOTE]
+> CPU resources are already over-committed in OpenShift today, thus the
+> limiting factor - usually - are memory resources.
+
+### User Stories
+
+* As an administrator, I want to be able to enable higher workload
+  density for OpenShift Virtualization clusters so that I can increase
+  the memory utilization.
+
+### Goals
+
+#### Timeline
+
+* GA higher workload density in OpenShift Virtualization in 2024
+
+#### Functionality
+
+* Fit more virtual machines onto a node once higher workload density
+  is enabled
+* Integrate well with [KSM] and [FPR]
+* **Technology Preview** - Enable higher density at all, limited
+  support for stressed clusters
+* **General Availability** - Improve handling of stressed clusters
+
+#### Usability
+
+* Provide a boolean in the OpenShift Console for enabling higher-density.
+  Do not require any additional workload or cluster level configuration
+  besides the initial provisioning.
+
+### Non-Goals
+
+* Complete life-cycling of the WASP Agent. We are not intending to write
+  an Operator for memory over commit for two reasons:
+  * [Kubernetes SWAP] is close, writing a fully fledged operator seems
+    to be no good use of resources
+  * To simplify the transition from WASP to [Kubernetes SWAP]
+
+## Proposal
+
+A higher workload density is achieved by combining two mechanisms
+
+1. Under-request memory resources for virtual machines by having the VM
+   pods (`launcher`) `requests.memory` being smaller than the VMs `guest.memory`
+   according to a configured ratio. This is owned by
+   [KubeVirt and present today](https://kubevirt.io/user-guide/operations/node_overcommit/).
```

Comment:
That's not 100% accurate, since in the non-overcomitted case:
`requests.memory == guest.memory + additional_infra_overhead`.

Overcommitment means that the memory requests become lower, but it doesn't necessarily mean that the new requests would be smaller than the defined guest memory.